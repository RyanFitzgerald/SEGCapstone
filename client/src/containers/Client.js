import React from 'react';
import { Switch, Route, Redirect } from 'react-router-dom';
import * as api from '../api';
import arraySort from 'array-sort';

// Import client components
import Submenu from '../components/clients/Submenu';
import Add from '../components/clients/Add';
import Edit from '../components/clients/Edit';
import Directory from '../components/clients/Directory';
import View from '../components/clients/View';
import Note from '../components/clients/Note';

class Client extends React.Component {
  constructor() {
    super();

    // Bind functions
    this.setActiveSubtab = this.setActiveSubtab.bind(this);
    this.getClients = this.getClients.bind(this);
    this.addToClients = this.addToClients.bind(this);
    this.updateClients = this.updateClients.bind(this);
    this.removeFromClients = this.removeFromClients.bind(this);
    this.renderError = this.renderError.bind(this);
    this.sortByKey = this.sortByKey.bind(this);
    
    // State
    this.state = {
      activeSubtab: 1,
      clients: null,     
      sort: {
        name: null,
        email: null,
        city: null
      }
    };
  }

  componentDidMount() {
    // Set Page Title
    document.title = 'Clients Overview | Renovaction';

    // Update tab
    this.props.setActiveTab(3);

    // Get clients
    this.getClients();
  }

  setActiveSubtab(tab) {
    this.setState({activeSubtab: tab});
  }

  getClients(query) {
    api.getClients(query).then(result => {
      const clients = result;
      this.setState({ clients });
    });
  }

  addToClients(client) {
    const clients = [...this.state.clients];
    clients[this.state.clients.length] = client;
    this.setState({ clients });
  }

  removeFromClients(id) {
    const clients = [...this.state.clients];
    const updated = clients.filter(el => {
      return el._id !== id;
    });
    this.setState({ clients: updated });
  }

  updateClients(client) {
    const clients = [...this.state.clients];
    const key = Object.keys(clients).find(key => clients[key]._id === client._id);
    clients[key] = client;
    this.setState({ clients });
  }

  sortByKey(array, key) {
    let asc = 'asc';
    let desc = 'desc'
    let sortOrder = {name: null, email: null, city: null};
    const arr = Object.keys(array).map((k) => array[k]);
    var sortedArray = [];

    if (key === 'name') {
      if(this.state.sort.name === asc) {
        sortOrder.name = desc;
        sortedArray = arraySort(arr, key, {reverse: true});
      }
      else {
        sortOrder.name = asc;
        sortedArray = arraySort(arr, key);
      }
    }
    else if (key === 'email') {
      if(this.state.sort.email === asc) {
        sortOrder.email = desc;
        sortedArray = arraySort(arr, key, {reverse: true});
      }
      else {
        sortOrder.email = asc;
        sortedArray = arraySort(arr, key);;
      }
    }
    else if (key === 'city') {
      if(this.state.sort.city === asc) {
        sortOrder.city = desc;
        sortedArray = arraySort(arr, key, {reverse: true});
      }
      else {
        sortOrder.city = asc;
        sortedArray = arraySort(arr, key);;
      }
    }
    
    this.setState({clients: sortedArray, sort: sortOrder});
  }

  renderError(formError) {
    if (!formError) return;
    return (
      <div className="flash flash--warn">
        <p>{formError}</p>
      </div>
    );
  }

  render() {
    const level = JSON.parse(sessionStorage.getItem('user')).role.level;
    return (
      <div>
        <Submenu activeSubtab={this.state.activeSubtab} level={this.props.level} checkLevel={this.props.checkLevel}/>

        <Switch>
          <Route exact path="/clients" render={() =>
            <Directory setActiveSubtab={this.setActiveSubtab} clients={this.state.clients} getClients={this.getClients} sortByKey={this.sortByKey} />
          }/>
          <Route path="/clients/add" render={() => (
            (level < 2) ? (
              <Redirect to='/'/>
            ) : (
              <Add setActiveSubtab={this.setActiveSubtab} addNotification={this.props.addNotification} renderError={this.renderError} addToClients={this.addToClients} level={this.props.level} checkLevel={this.props.checkLevel}/>
            )
          )}/>
          <Route path="/clients/:id/note" render={(location) => (
            (level < 2) ? (
              <Redirect to='/'/>
            ) : (
              <Note setActiveSubtab={this.setActiveSubtab} addNotification={this.props.addNotification} renderError={this.renderError} location={location} />
            )
          )}/>
          <Route path="/clients/:id/edit" render={(location) => (
            (level < 2) ? (
              <Redirect to='/'/>
            ) : (
              <Edit setActiveSubtab={this.setActiveSubtab} addNotification={this.props.addNotification} renderError={this.renderError} location={location} updateClients={this.updateClients}/>
            )
          )}/>
          <Route path="/clients/:id" render={(location) =>
            <View setActiveSubtab={this.setActiveSubtab} addNotification={this.props.addNotification} location={location} removeFromClients={this.removeFromClients}/>
          }/>
        </Switch>
      </div>
    );
  }
}

export default Client;