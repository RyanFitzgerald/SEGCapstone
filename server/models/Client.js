/*
  Client Model
  - This file contains the model for clients
*/
