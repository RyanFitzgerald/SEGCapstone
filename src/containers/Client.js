import React, { Component, PropTypes } from 'react';
import {withRouter, Route, Switch, Link} from 'react-router-dom';
import * as api from '../api';

// Import Client Components
import Summary from '../components/clients/Summary';
import Add from '../components/clients/Add';
import Edit from '../components/clients/Edit';
import View from '../components/clients/View';
import Search from '../components/clients/Search';

class Client extends Component {

    constructor(props) {
        super(props);

        // States
        this.state = {
            headerTab: 1,
            cities: null,
            clients: null,
            lastUpdated: null
        };
    }

    setHeaderTab = (tab) => {
        this.setState({
            headerTab: tab
        });
    }

    componentDidMount() {
        // Update page title
        document.title = 'Clients Overview';

        // Update Active Tab
        this.props.setActiveTab(3);

        // Update Header Title
        this.props.setHeaderTitle('Clients');

        // Get cities
        this.getCities();

        // Get clients
        this.getClients();
    }

    addClient = client => {
        api.addClient(client).then(resp => {
            this.getClients();
            this.props.history.push(`/clients/${resp.clientID}`);
        });
    }

    editClient = client => {
        api.editClient(client).then(resp => {
            this.getClients();
            this.props.history.push(`/clients/${resp.clientID}`);
        });
    }

    deleteClient = clientID => {
        api.deleteClient(clientID).then(() => {
            this.getClients();
            this.props.history.push('/clients');
        });
    }

    getCities = () => {
        api.getCities().then(cities => {
            this.setState({
                cities
            });
        });
    }

    getClients = () => {
        api.getClients().then(clients => {
            this.setState({
                clients
            });
        });
    }

    render() {
        return (
            <div className="clients">
                <div id="page-navigation">
                    <ul>
                        <li className={(this.state.headerTab == 1) ? 'active' : ''}>
                            <Link to="/clients">Overview</Link>
                        </li>
                        <li className={(this.state.headerTab == 2) ? 'active' : ''}>
                            <Link to="/clients/search">Search</Link>
                        </li>
                        <li className={(this.state.headerTab == 3) ? 'active' : ''}>
                            <Link to="/clients/add">Add</Link>
                        </li>
                    </ul>
                </div>
                {/* <!-- End page-navigation --> */}

                <Switch>
                    <Route exact path="/clients" render={() =>
                        <Summary setHeaderTab={this.setHeaderTab} />
                    } />
                    <Route path="/clients/search" render={() =>
                        <Search setHeaderTab={this.setHeaderTab} clients={this.state.clients} />
                    } />
                    <Route path="/clients/add" render={() =>
                        <Add setHeaderTab={this.setHeaderTab} addClient={this.addClient} cities={this.state.cities} />
                    } />
                    <Route path="/clients/:clientID/edit" render={({match}) =>
                        <Edit setHeaderTab={this.setHeaderTab} match={match} editClient={this.editClient} cities={this.state.cities} />
                    } />
                    <Route path="/clients/:clientID" render={({match}) =>
                        <View setHeaderTab={this.setHeaderTab} match={match} deleteClient={this.deleteClient} />
                    } />
                </Switch>
            </div>
        );
    }
}

Client.propTypes = {
    setActiveTab: PropTypes.func.isRequired,
    setHeaderTitle: PropTypes.func.isRequired,
    history: PropTypes.any
};

export default withRouter(Client);
