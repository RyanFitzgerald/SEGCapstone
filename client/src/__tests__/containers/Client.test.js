import React from 'react';
import Client from '../../Containers/Client';

describe('<Client />', () => {
  beforeAll(() => {
    // Add necessary info to user
    const user = {
      role: {
        level: 1
      }
    };
    sessionStorage.setItem('user', JSON.stringify(user));
  });

  it('should render correctly', () => {
    // Return empty objects on requests
    mock.onAny().reply(200, {});

    // Create mock functions / spies
    const setActiveTabMock = jest.fn();
    const getClientsSpy = jest.spyOn(Client.prototype, 'getClients');

    // Shallow render it
    const wrapper = shallow(
      <Client setActiveTab={setActiveTabMock}/>
    );

    // Compare to old snapshot
    expect(shallowToJson(wrapper)).toMatchSnapshot();

    // Check set active tab is called and with correct value
    expect(setActiveTabMock).toHaveBeenCalledTimes(1);
    expect(setActiveTabMock.mock.calls[0][0]).toBe(3);

    // Check if component mounting methods were called
    expect(getClientsSpy).toHaveBeenCalled();
  });
});

