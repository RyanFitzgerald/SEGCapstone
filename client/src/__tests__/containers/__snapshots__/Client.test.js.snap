// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Client /> should render correctly 1`] = `
<div>
  <Submenu
    activeSubtab={1}
  />
  <Switch>
    <Route
      exact={true}
      path="/clients"
      render={[Function]}
    />
    <Route
      path="/clients/add"
      render={[Function]}
    />
    <Route
      path="/clients/:id/note"
      render={[Function]}
    />
    <Route
      path="/clients/:id/edit"
      render={[Function]}
    />
    <Route
      path="/clients/:id"
      render={[Function]}
    />
  </Switch>
</div>
`;
